pkg/ruler/ruler_test.go

```diff
@@ -216,3 +227,324 @@ func compareRuleGroupDescToStateDesc(t *testing.T, expected *rules.RuleGroupDesc
+			},
+
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, unhealthy ACTIVE ruler": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.ACTIVE)
+				desc.Ingesters[ruler2] = ring.IngesterDesc{
+					Addr:      ruler2Addr,
+					Timestamp: time.Now().Add(-time.Hour).Unix(),
+					State:     ring.ACTIVE,
+					Tokens:    []uint32{user1Group2Token + 1, user3Group1Token + 1},
+				}
+			},
+
+			// This ruler doesn't rulers from unhealthy ruler (RF=1).
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, LEAVING ruler-1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.LEAVING)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.ACTIVE)
+			},
+
+			// LEAVING still has its rules.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, LEAVING ruler-2": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.LEAVING)
+			},
+
+			// No extra rules from LEAVING.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
+				user1: {user1Group1},
+				user2: {user2Group1},
+			},
+		},
+
+		"default sharding, JOINING ruler-1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.JOINING)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.ACTIVE)
+			},
+
+			// JOINING ruler has no rules yet.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{},
+		},
+
+		"default sharding, JOINING ruler-2": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", []uint32{user1Group1Token + 1, user2Group1Token + 1}, ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", []uint32{user1Group2Token + 1, user3Group1Token + 1}, ring.JOINING)
+			},
+
+			// JOINING ruler will get the rules.
+			expectedRulesForFuler1: map[string]rules.RuleGroupList{
```

Comment:
Doesn't this expectation conflicts with `default sharding, JOINING ruler-1`?

Reply:
Right. The comment is wrong. Those rules will in fact not be given to any ruler!

Reply:
This is result of using replication factor 1 :( With higher RF, only one ruler would be joining, other would be ACTIVE or LEAVING and they would get the rule instead. With RF=1, we only select single ruler.

We could fix this by extending the replica set if JOINING ingester was selected.

Reply:
Ruler now uses different replication strategy. Only ACTIVE Rulers can receive rules. JOINING/LEAVING/PENDING rulers have no rules.